include/deal.II/matrix_free/tools.h

```diff
@@ -98,6 +98,64 @@ namespace MatrixFreeTools
     const unsigned int quad_no                  = 0,
     const unsigned int first_selected_component = 0);
 
+  /**
+   * Compute the diagonal of a linear operator (@p diagonal_global), given
+   * @p matrix_free and the local cell integral operation @p local_vmult. The
+   * vector is initialized to the right size in the function.
+   *
+   * In contrast to the function above, this function works on a vector of
+   * FEEvaluation objects with single component. This is particularly useful if
+   * the number of components is not known at compile time.
+   */
+  template <int dim,
+            int fe_degree,
+            int n_q_points_1d,
+            typename Number,
+            typename VectorizedArrayType,
+            typename VectorType>
+  void
+  compute_diagonal(
+    const MatrixFree<dim, Number, VectorizedArrayType> &matrix_free,
+    VectorType &                                        diagonal_global,
+    const int                                           n_components,
+    const std::function<void(std::vector<FEEvaluation<dim,
+                                                      fe_degree,
+                                                      n_q_points_1d,
+                                                      1,
+                                                      Number,
+                                                      VectorizedArrayType>> &)>
```

Comment:
I assume the main reason you spell out all template parameters here, rather than making this function a `template <typename FEEvalType>`, is to get better error messages at compile time, plus the documentation requirement in terms of what a user should expect to deliver as an input argument? (Not that I disagree here, but I was confused about why all these template parameters needed to be listed in the first place, and tried to find the main reasons.)

Reply:
> I assume the main reason you spell out all template parameters here, rather than making this function a template <typename FEEvalType>, is to get better error messages at compile time, plus the documentation requirement in terms of what a user should expect to deliver as an input argument? 

Yes, it is better to be explicit. Else one could plug in `FEFaceEvaluation`, ...